PURL-TYPES.rst

```diff
@@ -521,6 +521,44 @@ swift
       pkg:swift/github.com/Alamofire/Alamofire@5.4.3
       pkg:swift/github.com/RxSwiftCommunity/RxFlow@2.12.4
 
+vcpkg
+-----
+``vcpkg`` for `vcpkg <https://vcpkg.io>`_ C and C++ packages:
+
+- By default, vcpkg will build C and C++ dependencies from source, using configurations compatible with the user's project. For this and other reasons, it uses different terminology, documented here.
+
+  - **port** - A package, along with it's build scripts and possibly minor modifications.
+  - **registry** - A collection of ports, possibly private to the user. Analogous to **repository**.
+
+- ``namespace``: Currently reserved for future use and should be empty
+
+- ``name``: The case-sensitive name of the port. **Required**
+- ``version``: The port version. **Required**
+- ``qualifiers``: The qualifiers below are used to provide more specific information on a port's origin registry and to distinguish between multiple instances of the same port dependency within the same project (for example, when the project targets multiple platforms).
+
+  - ``repository_url``: A `percent-encoded <https://www.rfc-editor.org/rfc/rfc3986#section-2.1>`_ absolute URI for the registry the port came from. For filesystem registries, the URI will have a `file` URI scheme. **Optional**
+
+    - If empty, the port comes from the default registry, ``https://github.com/microsoft/vcpkg``.
+
+  - ``registry_version``: A string specifying the version of the registry specified by `repository_url`. 
```

Comment:
Does the registry require immutability of package names and versions? If so, it seems like you could make this optional and improve the legibility of the purls that flow from it.

Reply:
Short answer - No, I don't think they are immutable.

Longer answer - @ras0219-msft or @BillyONeal might provide more clarity, but as I understand it, each port has a version for the package, but also has a distinct port version that should be increased any time the port changes, which happens when changes are made that are *not* a package version update. You can see this behavior in https://github.com/microsoft/vcpkg/pull/33333/files#diff-dba19656e5032ab9d4d3f06eea74404e06a570a1552e628ab1410b721b2481ef and https://github.com/microsoft/vcpkg/pull/33351/files#diff-dba19656e5032ab9d4d3f06eea74404e06a570a1552e628ab1410b721b2481ef. It is also possible for a port/package to be removed, although it would show up in the history of the registry (since they are usually implemented via git repositories).  I think that this means that you need the package version + one of either the port version or the registry version (commit hash). With the package version, the port version should be sufficient to identify a package in a registry, but not sufficient to trace, as the registry might have removed it (or some other sort of deviation, since private registries are not under the control of the vcpkg team). Whereas the registry version should be sufficient for all of those, as long as git history doesn't get rewritten.

Reply:
This has now been marked as optional. The `port_revision` and version and package name should together be unique already. However, the `registry_revision` may still be specified as an optional qualifier to add traceability (e.g. it may be that somebody edits the port but forgets to update the port revision).

Reply:
The risk that a mistake with updating a port revision is higher in private registries where a maintainer may not have as strict auditing/review practices. This makes it so that automated tools that produce PURLs can provide helpful tracing back to the exact commit of the recipe that was used (as opposed to doing version comparisons to test for vulnerable versions).